rosbag2_cpp/src/rosbag2_cpp/reindexer.cpp

```diff
@@ -0,0 +1,252 @@
+namespace rosbag2_cpp
+{
+Reindexer::Reindexer(
+  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory,
+  std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io)
+: storage_factory_(std::move(storage_factory)),
+  metadata_io_(std::move(metadata_io)) {}
+
+Reindexer::~Reindexer() {}
+
+bool Reindexer::compare_relative_file(
+  const rcpputils::fs::path & first_path, const rcpputils::fs::path & second_path)
+{
+  std::regex regex_rule(".*_(\\d+)\\.db3", std::regex_constants::ECMAScript);
+
+  std::smatch first_match;
+  std::smatch second_match;
+
+  auto first_path_string = first_path.string();
+  auto second_path_string = second_path.string();
+
+  auto first_regex_good = std::regex_match(first_path_string, first_match, regex_rule);
+  auto second_regex_good = std::regex_match(second_path_string, second_match, regex_rule);
+
+  // Make sure the paths have regex matches
+  if (!first_regex_good || !second_regex_good) {
+    throw std::runtime_error("Malformed relative file name. Expected numerical identifier.");
+  }
+
+  u_int32_t first_db_num = std::stoul(first_match.str(1), nullptr, 10);
+  u_int32_t second_db_num = std::stoul(second_match.str(1), nullptr, 10);
+
+  return first_db_num < second_db_num;
+}
+
+std::vector<rcpputils::fs::path> Reindexer::get_bag_files(
+  const rcpputils::fs::path & base_folder)
+{
+  // Look in the uri directory to see what database files are there
+  std::vector<rcpputils::fs::path> output;
+
+  auto allocator = rcutils_get_default_allocator();
+  auto dir_iter = rcutils_dir_iter_start(base_folder.string().c_str(), allocator);
+
+  // Make sure there are files in the directory
+  if (dir_iter == nullptr) {
+    throw std::runtime_error("Empty directory.");
+  }
+
+  // Get all file names in directory
+  do {
+    auto found_file = rcpputils::fs::path(dir_iter->entry_name);
+
+    // We are ONLY interested in database files
+    if (found_file.extension().string() == ".db3") {
+      auto full_path = base_folder / found_file;
+      output.emplace_back(full_path);
+    }
+  } while (rcutils_dir_iter_next(dir_iter));
+
+  // Sort relative file path by database number
+  std::sort(
+    output.begin(), output.end(),
+    [](rcpputils::fs::path a, rcpputils::fs::path b) {return compare_relative_file(a, b);});
+
+  return output;
+}
+
+void Reindexer::fill_topics_metadata()
+{
+  // rcpputils::check_true(storage_ != nullptr, "Bag is not open. Call open() before reading.");
+  topics_metadata_.clear();
+  topics_metadata_.reserve(metadata_.topics_with_message_count.size());
+  for (const auto & topic_information : metadata_.topics_with_message_count) {
+    topics_metadata_.push_back(topic_information.topic_metadata);
+  }
+}
+
+void Reindexer::init_metadata(
+  const std::vector<rcpputils::fs::path> & files,
+  const rosbag2_storage::StorageOptions & storage_options)
+{
+  metadata_ = rosbag2_storage::BagMetadata{};
+
+  // This reindexer will only work on SQLite files, so this can't change
```

Comment:
I don't think this comment makes sense. We expect that all bags use the same storage implementation - that's all.

Reply:
It was an old comment that I kept overlooking. It's gone now.